### PR TITLE
Music: use shared busses for expensive sound effects

### PR DESCRIPTION
**Pull request by breville**
With this change, the filter and delay sound effects are no longer created separately for each sound that's played.  Rather, we pre-generate "busses" for each possible combination of values, and when a sound is played, it is attached to one of these existing busses.  We continue to create a volume effect for each sound that's played.

### Details

Screenshot 2023-03-15 at 1 05 42 AM

This morning, doing a quick demo for @simonguest, we noticed that an opening sound (the bass) wasn't always playing.  This was happening because a bit of time had passed between starting to play the song, and the time that we actually tried to play the bass, and we didn't have any grace period.  We do have a grade period of `0.1` seconds for a triggered sound, and so one mitigation is to add a grade period for non-triggered sounds as well.  `0.05` has been added, and we'll see how that goes on various systems.  Of course, this grace period isn't ideal, as the longer it takes to get these opening sounds playing, the more out of sync they will be with each other, and indeed the rest of the song.  

As @sanchitmalhotra126 and I discussed this more, we started to wonder why it would take much time to simply play the sounds and add them to an array.  (That said, it was 384 sounds in `drums` and 64 more in `bass1`.)  One big realisation was that, particularly as our projects get more complex with drums and keyboard notes, we probably don't want to be creating unique effects for each sound, as done in https://github.com/code-dot-org/code-dot-org/pull/50356.  Rather, we decided it makes sense for the `filter` and `delay` effects, at least, to run on a shared "bus" that any sound can route into.  We figured that volume should be lightweight enough that it can still be used per sound, potentially allowing us to give a greater variety of values.  On the other hand, we can potentially offer some global controls over parameters for the `filter` and `delay` busses.

This change sets up a set of shared busses for each potential combination of the `filter` and `delay` effects and their values of `normal` (none), `medium`, and `low`.  (Setting up these combinations is another reason to keep volume per sound, to stop the set from getting too large.)

Just using the audio clock for some rudimentary profiling, it appears that we have roughly halved the time to schedule all the sound events in the test song.

### CODE REVIEW

File: apps/src/music/player/soundEffects.js [hunk 1]
```diff
@@ -1,62 +1,165 @@
...
+    if (effects.filter === 'low') {
+      keyStrings.push('filter_low');
+    }
+    if (effects.delay === 'medium') {
+      keyStrings.push('delay_medium');
+    }
+    if (effects.delay === 'low') {
+      keyStrings.push('delay_low');
+    }
+    const keyString = keyStrings.join('-');
```
Comment (sanchitmalhotra126) [inline apps/src/music/player/soundEffects.js]: minor: could move this `effect` -> `keyString` conversion into a separate function to reuse here and in `generateBusses()`?

File: apps/src/music/player/soundEffects.js [hunk 2]
```diff
@@ -1,62 +1,165 @@
...
+  generateBus(effects) {
+    let firstNode = null;
+    let lastNode = null;
+
+    if (['low', 'medium'].includes(effects.filter)) {
+      const filter = this.audioContext.createBiquadFilter();
+      filter.type = 'lowpass';
+      filter.frequency.value = effects.filter === 'low' ? 800 : 3000;
+
+      if (lastNode) {
```
Comment (sanchitmalhotra126) [inline apps/src/music/player/soundEffects.js]: Will `lastNode` ever be defined? Looks like it's always `null` at the start of this function?

**Review by sanchitmalhotra126 — APPROVED** [review]
Sweet!
> it appears that we have roughly halved the time to schedule all the sound events in the test song.

🎉 🎉 🎉

File: apps/src/music/player/soundEffects.js (hunk 1 above)
Comment (breville) [reply apps/src/music/player/soundEffects.js]: Great call.  Done.

File: apps/src/music/player/soundEffects.js (hunk 2 above)
Comment (breville) [reply apps/src/music/player/soundEffects.js]: It won't.  I had done this deliberately to create a kind of duplicatable template, but it really is unreachable code, so I'll remove it.

**Pull request merged**